posix/unistd.c

```diff
@@ -0,0 +1,612 @@
+    char filepath[OE_PATH_MAX] = {0};
+    char newfilepath[OE_PATH_MAX] = {0};
+
+    if (!(fs = oe_mount_resolve(oldpath, filepath)))
+        OE_RAISE_ERRNO(oe_errno);
+
+    if (!(newfs = oe_mount_resolve(newpath, newfilepath)))
+        OE_RAISE_ERRNO(oe_errno);
+
+    if (fs != newfs)
+        OE_RAISE_ERRNO(OE_EXDEV);
+
+    ret = fs->ops.fs.link(fs, filepath, newfilepath);
+
+done:
+    return ret;
+}
+
+int oe_link_d(uint64_t devid, const char* oldpath, const char* newpath)
+{
+    int ret = -1;
+
+    if (devid == OE_DEVID_NONE)
+    {
+        ret = oe_link(oldpath, newpath);
+    }
+    else
+    {
+        oe_device_t* dev;
+
+        if (!(dev = oe_device_table_get(devid, OE_DEVICE_TYPE_FILE_SYSTEM)))
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        ret = dev->ops.fs.link(dev, oldpath, newpath);
+    }
+
+done:
+    return ret;
+}
+
+int oe_unlink(const char* pathname)
+{
+    int ret = -1;
+    oe_device_t* fs = NULL;
+    char filepath[OE_PATH_MAX] = {0};
+
+    if (!(fs = oe_mount_resolve(pathname, filepath)))
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = fs->ops.fs.unlink(fs, filepath);
+
+done:
+    return ret;
+}
+
+int oe_unlink_d(uint64_t devid, const char* pathname)
+{
+    int ret = -1;
+
+    if (devid == OE_DEVID_NONE)
+    {
+        ret = oe_unlink(pathname);
+    }
+    else
+    {
+        oe_device_t* dev;
+
+        if (!(dev = oe_device_table_get(devid, OE_DEVICE_TYPE_FILE_SYSTEM)))
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        ret = dev->ops.fs.unlink(dev, pathname);
+    }
+
+done:
+    return ret;
+}
+
+int oe_truncate(const char* pathname, oe_off_t length)
+{
+    int ret = -1;
+    oe_device_t* fs = NULL;
+    char filepath[OE_PATH_MAX] = {0};
+
+    if (!(fs = oe_mount_resolve(pathname, filepath)))
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = fs->ops.fs.truncate(fs, filepath, length);
+
+done:
+    return ret;
+}
+
+int oe_truncate_d(uint64_t devid, const char* path, oe_off_t length)
+{
+    int ret = -1;
+
+    if (devid == OE_DEVID_NONE)
+    {
+        ret = oe_truncate(path, length);
+    }
+    else
+    {
+        oe_device_t* dev;
+
+        if (!(dev = oe_device_table_get(devid, OE_DEVICE_TYPE_FILE_SYSTEM)))
+            OE_RAISE_ERRNO(OE_EINVAL);
+
+        ret = dev->ops.fs.truncate(dev, path, length);
+    }
+
+done:
+    return ret;
+}
+
+oe_off_t oe_lseek(int fd, oe_off_t offset, int whence)
+{
+    oe_off_t ret = -1;
+    oe_fd_t* file;
+
+    if (!(file = oe_fdtable_get(fd, OE_FD_TYPE_FILE)))
+        OE_RAISE_ERRNO(oe_errno);
+
+    ret = file->ops.file.lseek(file, offset, whence);
+
+done:
+    return ret;
+}
+
+ssize_t oe_readv(int fd, const struct oe_iovec* iov, int iovcnt)
+{
+    ssize_t ret = -1;
+    ssize_t nread = 0;
+
+    if (fd < 0 || !iov)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    for (int i = 0; i < iovcnt; i++)
+    {
+        const struct oe_iovec* p = &iov[i];
+        ssize_t n;
+
+        n = oe_read(fd, p->iov_base, p->iov_len);
```

Comment:
Doesn't this potentially break the atomic promise of readv/writev if we end up with ocalls on each vector? 

Also, it seems like an incredible performance overhead that would not be expected by a user of readv/writev. 

It seems like this needs to be a single call, with a data structure set up to support doing this from the host side.